src/server/db_slice.cc

```diff
@@ -816,42 +827,45 @@ OpResult<int64_t> DbSlice::UpdateExpire(const Context& cntx, PrimeIterator prime
   }
 }
 
-std::pair<PrimeIterator, bool> DbSlice::AddOrUpdateInternal(const Context& cntx,
-                                                            std::string_view key, PrimeValue obj,
-                                                            uint64_t expire_at_ms,
-                                                            bool force_update) noexcept(false) {
+DbSlice::AddOrFindResult DbSlice::AddOrUpdateInternal(const Context& cntx, std::string_view key,
+                                                      PrimeValue obj, uint64_t expire_at_ms,
+                                                      bool force_update) noexcept(false) {
   DCHECK(!obj.IsRef());
 
-  pair<PrimeIterator, bool> res = AddOrFind(cntx, key);
-  if (!res.second && !force_update)  // have not inserted.
+  auto res = AddOrFind(cntx, key);
```

Comment:
when you return res on this funciton the AutoUpdater will be invoked no?

Reply:
No...
`AutoUpdater` is a move-only type, i.e. it can't be copied.
When we return `res`, the `post_updater` member will be `move()`ed to the returned object. That `move()` will `Cancel()` the `AutoUpdater` within `res`.

Reply:
can you explicitly delete the copy constructor in AutoUpdater?

Reply:
Yeah, that's a good idea.

Reply:
It is implicitly `delete`d, but I'll make it explicit.